src/amms/balancer/mod.rs

```diff
@@ -0,0 +1,670 @@
+
+        let base_token_balance = self.liquidity[base_token_index];
+        let quote_token_balance = self.liquidity[quote_token_index];
+        let base_token_weight = self.weights[base_token_index];
+        let quote_token_weight = self.weights[quote_token_index];
+        let swap_fee = U256::from(self.fee);
+        let out = bmath::calculate_out_given_in(
+            base_token_balance,
+            base_token_weight,
+            quote_token_balance,
+            quote_token_weight,
+            amount_in,
+            swap_fee,
+        )
+        .map_err(BalancerError::from)?;
+        self.liquidity[base_token_index] = bmath::badd(base_token_balance, amount_in)?;
+        self.liquidity[quote_token_index] = bmath::bsub(quote_token_balance, out)?;
+        Ok(out)
+    }
+
+    async fn init<T, N, P>(
+        mut self,
+        block_number: BlockId,
+        provider: Arc<P>,
+    ) -> Result<Self, AMMError>
+    where
+        T: Transport + Clone,
+        N: Network,
+        P: Provider<T, N>,
+    {
+        let deployer =
+            IGetBalancerPoolDataBatchRequest::deploy_builder(provider, vec![self.address]);
+        let res = deployer.block(block_number).call_raw().await?;
+
+        let mut data =
+            <Vec<(Vec<Address>, Vec<u16>, Vec<U256>, Vec<U256>, u32)> as SolValue>::abi_decode(
+                &res, false,
+            )?;
+        let (tokens, decimals, liquidity, weights, fee) = if !data.is_empty() {
+            data.remove(0)
+        } else {
+            return Err(BalancerError::InitializationError.into());
+        };
+
+        self.tokens = tokens;
+        self.decimals = decimals.into_iter().map(|d| d as u8).collect();
+        self.liquidity = liquidity;
+        self.weights = weights;
+        self.fee = fee;
+
+        Ok(self)
+    }
+}
+
+impl BalancerPool {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        address: Address,
+        tokens: Vec<Address>,
+        decimals: Vec<u8>,
+        liquidity: Vec<U256>,
+        weights: Vec<U256>,
+        fee: u32,
+    ) -> BalancerPool {
+        BalancerPool {
+            address,
+            tokens,
+            decimals,
+            liquidity,
+            weights,
+            fee,
+        }
+    }
+
+    /// Updates the AMM data from a log.
+    #[instrument(skip(self), level = "debug")]
+    fn sync_from_log(&mut self, log: Log) -> Result<(), AMMError> {
+        let signature = log.topics()[0];
+
+        if IBPool::LOG_SWAP::SIGNATURE_HASH == signature {
+            self.sync_from_swap_log(log)?;
+        } else {
+            return Err(AMMError::UnrecognizedEventSignature(signature));
+        }
+
+        Ok(())
+    }
+
+    pub fn sync_from_swap_log(
+        &mut self,
+        log: Log,
+    ) -> Result<alloy::primitives::Log<IBPool::LOG_SWAP>, AMMError> {
+        let swap_event = IBPool::LOG_SWAP::decode_log(log.as_ref(), true)?;
```

Comment:
Lets get rid of these functions and move the logic directly into the `sync()` impl.